### PR TITLE
adds on discard support for `FluxZip`/`MonoZip`

### PR DESCRIPTION
**Pull request by OlegDokuka**
This PR adds support for missing the `onDiscard` functionality to the `FluxZip` / `MonoZip` operators

Also, this PR includes fixes to fluky tests

Signed-off-by: Oleh Dokuka <odokuka@vmware.com>

### CODE REVIEW

File: reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java
Comment (chemicL) [inline reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java:460]: Is there no need to use the `onUndelivered` value like in `DefaultTestPublisher`?

File: reactor-core/src/main/java/reactor/core/publisher/FluxZip.java
Comment (OlegDokuka) [reply reactor-core/src/main/java/reactor/core/publisher/FluxZip.java:717]: This is already done inside onDiscard

File: reactor-core/src/main/java/reactor/core/publisher/FluxZip.java
Comment (OlegDokuka) [reply reactor-core/src/main/java/reactor/core/publisher/FluxZip.java:699]: As of the fusion contract, the call for the `queue.clear` means - termination of the process. In case of fusion - queue comes from the upstream, so the upstream is a creator / owner of the queue, so it and only it can have full control of when values are offered and drained. See https://github.com/reactor/reactor-core/issues/2636 for more information

File: reactor-core/src/test/java/reactor/core/publisher/OnDiscardShouldNotLeakTest.java
Comment (OlegDokuka) [reply reactor-core/src/test/java/reactor/core/publisher/OnDiscardShouldNotLeakTest.java:107]: Scenarios are defined statically, so within a call, we can not get access to the allocator which holds all the references to `Tracked` objects. In this particular case, we initiate a specific path for `FluxZip`, which uses `ZipSingleCoordinator` instead of the default `ZipCoordinator`. Thus, it is not that important for now, although this can be improved in a separate PR so the rules API can get access to allocator upon the rule installation

File: reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java
Comment (OlegDokuka) [reply reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java]: I rolled back this API since it is a breaking one after all. In the case of `ColdTestPublisher`, all the values are cached in the internal store, so they can be consumed slightly latter

File: reactor-core/src/main/java/reactor/core/publisher/FluxZip.java
Comment (chemicL) [reply reactor-core/src/main/java/reactor/core/publisher/FluxZip.java:699]: thanks, that's what I was asking about.

File: reactor-core/src/main/java/reactor/core/publisher/FluxZip.java
Comment (chemicL) [reply reactor-core/src/main/java/reactor/core/publisher/FluxZip.java:717]: That is what I am saying, but without the "already" part. My suggestion is to do the null check and fail early without executing the following 4 if statements + 1 method call, all leading to a no-op.

File: reactor-core/src/main/java/reactor/core/publisher/MonoZip.java
Comment (chemicL) [inline reactor-core/src/main/java/reactor/core/publisher/MonoZip.java:261]:
```suggestion
			if (DONE.getAndSet(this, Integer.MIN_VALUE) <= 0) {
```

File: reactor-core/src/test/java/reactor/core/publisher/OnDiscardShouldNotLeakTest.java
Comment (chemicL) [reply reactor-core/src/test/java/reactor/core/publisher/OnDiscardShouldNotLeakTest.java:107]: I see. Now I understand it's not vital to validate the scalar value is released. Thanks.

File: reactor-test/src/main/java/reactor/test/publisher/TestPublisher.java
Comment (chemicL) [inline reactor-test/src/main/java/reactor/test/publisher/TestPublisher.java:21]: As the API got reverted, the unused imports can be removed.

**Review by chemicL — CHANGES_REQUESTED** [review]
I have one doubt that seems significant, about discarding items in MonoZip, but everything else looks good. Happy to approve once I understand the discarding in MonoZip.

File: reactor-core/src/main/java/reactor/core/publisher/MonoZip.java
Comment (OlegDokuka) [reply reactor-core/src/main/java/reactor/core/publisher/MonoZip.java:350]: generally, this should never happen as of Mono specification which guarantees only the following cases: 

`onNext + onComplete`
`onComplete`
`onError`

The others are prohibited so onNext + onError was never expected although I added that as an extra check. 

Will roll it back

File: reactor-core/src/main/java/reactor/core/publisher/MonoZip.java
Comment (OlegDokuka) [reply reactor-core/src/main/java/reactor/core/publisher/MonoZip.java:342]: false means the signal was not delivered which happens when the process was canceled or force terminated with the error or empty source. In that scenario value is discarded by a **caller** since the `signal` method has no information about the **caller** or signaled **value**

Comment (chemicL) [reply reactor-core/src/main/java/reactor/core/publisher/MonoZip.java:342]: The way it is implemented right now seems different. Also, please have a look at failing tests from CI.
The condition for returning false is:

            if (DONE.incrementAndGet(this) != n) {
 				return false;
 			}

And that is the same condition which signifies the zipping is done. I'm a bit confused.

File: reactor-core/src/main/java/reactor/core/publisher/MonoZip.java
Comment (OlegDokuka) [reply reactor-core/src/main/java/reactor/core/publisher/MonoZip.java]: fixed

File: reactor-core/src/main/java/reactor/core/publisher/MonoZip.java
Comment (OlegDokuka) [reply reactor-core/src/main/java/reactor/core/publisher/MonoZip.java:342]: right

File: reactor-test/src/main/java/reactor/test/publisher/TestPublisher.java
Comment (OlegDokuka) [reply reactor-test/src/main/java/reactor/test/publisher/TestPublisher.java]: done

File: reactor-core/src/main/java/reactor/core/publisher/MonoZip.java
Comment (OlegDokuka) [reply reactor-core/src/main/java/reactor/core/publisher/MonoZip.java:347]: fixed

File: reactor-core/src/test/java/reactor/core/publisher/MonoCacheInvalidateWhenTest.java
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/MonoCacheInvalidateWhenTest.java:268]: `assumeThat` is to skip a test under a condition that is evaluated at runtime... I wouldn't use that here. maybe a log if the number of "lessInteresting" cases goes beyond `loops / 2` instead of an assertion?

it sounds like the goal / expectation was to trigger at least 50% of interesting cases ? maybe we can just check that `lessInteresting` `isBetween(0, loops - 1)` instead? (at least 1 interesting case asserted).

File: reactor-core/src/test/java/reactor/core/publisher/FluxFilterFuseableTest.java
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/FluxFilterFuseableTest.java:286]: does this change really relates to this PR?

File: reactor-core/src/test/java/reactor/core/publisher/FluxFilterFuseableTest.java
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/FluxFilterFuseableTest.java:169]: does this change really relates to this PR?

**Review by simonbasle — CHANGES_REQUESTED** [review]
a few comments on the tests, plus minor note on change from onNextDropped to onDiscard

File: reactor-core/src/main/java/reactor/core/publisher/FluxZip.java
Comment (OlegDokuka) [reply reactor-core/src/main/java/reactor/core/publisher/FluxZip.java:534]: `isBetween` checks that number of events is between 0 and 50%. Actually - number can be 0 as well as 100% since verification relies on technically random event with **finit** number of observations. 

If we puts that on clear statistic - both events have equal probability. However, that is true for **infinit** number of experiments. If we have **fixed** number of experiments, we have 0 guarantees to get exactly 50 as well no guarantee we always see less or more.

Thus, this test from statistical verification is invalid and gives nothing.

Comment (OlegDokuka) [reply reactor-core/src/main/java/reactor/core/publisher/FluxZip.java:534]: Done in that case brings wrong impression. We have to remeber that upstream is flux and we set it as done once first element lands in the method. The other onNext call which may appear after cencellation must be discarded. Although i got your point. Will improve

File: reactor-core/src/test/java/reactor/core/publisher/MonoCacheInvalidateWhenTest.java
Comment (OlegDokuka) [reply reactor-core/src/test/java/reactor/core/publisher/MonoCacheInvalidateWhenTest.java:268]: `isBetween` checks that number of events is between 0 and 50%. In reality - that number can be 0 as well as 100% since verification relies on random event but with finit number of observations.

If we put that on clear statistic - both events have equal probability. However, that is true for infinit number of experiments. If we have fixed number of experiments, we have 0 guarantees to get exactly 50 as well no guarantee we always see less or more.

Thus, this test from statistical verification is invalid and gives nothing.

File: reactor-core/src/test/java/reactor/core/publisher/FluxFilterFuseableTest.java
Comment (OlegDokuka) [reply reactor-core/src/test/java/reactor/core/publisher/FluxFilterFuseableTest.java:286]: This fails on my machine all the time (m1 does goos job in detecting such things)

File: reactor-core/src/test/java/reactor/core/publisher/MonoCacheInvalidateWhenTest.java
Comment (simonbasle) [reply reactor-core/src/test/java/reactor/core/publisher/MonoCacheInvalidateWhenTest.java:268]: ok. I think as long as the `assumeThat` is the last statement the test should report any meaningful failure. it will just appear as skipped if it found no failure but couldn't make sure that interesting state was reached and asserted half the time

**Review by simonbasle — APPROVED** [review]
good from my perspective. @chemicL did you have any unresolved comment?

**Review by chemicL — APPROVED** [review]
LGTM. I had one comment about handling `dataSignal == null` case early in `FluxZip::drain` but now the logic is different. Anyways, it's not a blocker. Worth considering an optimisation to avoid a few branched executions just to hit a null check.

**Pull request merged**
